hilti/toolchain/src/compiler/optimizer.cc

```diff
@@ -1659,19 +1690,333 @@ struct DeadCodeVisitor : OptimizerVisitor, visitor::PreOrder<bool, DeadCodeVisit
+                        return {};
+                }
+
+                // FIXME(bbannier): If we construct a struct we can trigger life-cycle
+                // hooks making such functions not pure. Reject them here.
+
+                // Accessing keywords is side-effect free.
+                else if ( const auto& e = decl.tryAs<declaration::Expression>();
+                          e && e->expression().isA<expression::Keyword>() ) {
+                    continue;
+                }
+
+                else
+                    // Do not mark this function pure if it references any other kinds of IDs.
+                    return {};
+            }
+
+            else if ( auto call = child.node.tryAs<operator_::struct_::MemberCall>() ) {
+                const auto& op0 = call->op0();
+
+                const auto& callee = op0.type().tryAs<type::Struct>();
+                if ( ! callee )
+                    return {};
+
+                auto typeID = op0.type().typeID();
+                if ( ! typeID )
+                    return {};
+
+                const auto& method = call->op1().tryAs<expression::Member>();
+                if ( ! method )
+                    return {};
+
+                auto field = callee->field(method->id());
+                if ( ! field )
+                    return {};
+
+
+                // Method declared inline.
+                if ( const auto& functions = field->childRefsOfType<Function>(); ! functions.empty() ) {
+                    if ( const auto& attrs = functions[0]->as<Function>().attributes();
+                         ! attrs || ! attrs->find("&pure") )
+                        // Called method not marked pure.
+                        return {};
+                }
+
+                // Method not declared inline.
+                else if ( const auto& function =
+                              scope::lookupID<declaration::Function>(ID(util::fmt("%s::%s", *typeID, field->id())), p,
+                                                                     "function");
+                          function ) {
+                    if ( const auto& attrs = function->first->as<declaration::Function>().function().attributes();
+                         ! attrs || ! attrs->find("&pure") )
+                        // Called method not marked pure.
+                        return {};
+                }
+                else {
+                    return {};
+                }
+            }
+
+            else if ( child.node.isA<operator_::struct_::MemberNonConst>() )
+                return {};
+        }
+
+        auto new_attrs = AttributeSet({Attribute("&pure")});
+        if ( auto attrs = x.attributes() )
+            for ( const auto& attr : attrs->attributes() )
+                new_attrs = AttributeSet::add(std::move(new_attrs), attr);
+
+        return Function(x.id(), x.type(), x.body(), x.callingConvention(), new_attrs, x.meta());
```

Comment:
This becomes slightly tricky since the `position_t` we inject here and which we would edit does not necessarily point to a `Function`, but maybe also a `declaration::Function` if we visited a declaration. The code for field visitation also needs to deal with either inline method (`Function`) or out-of-line methods (`declaration::Function`). I'd prefer to leave the code here as is instead of trying to make that all align.

Reply:
What worries me here that we're potentially recreating a lot functions inside the AST, which aside from performance, can also make things confusing when comparing before/after ASTs. I think you could pass a `NodeRef` to the `Function` itself into `pureFunction` and then modify the function in place. If you currently can't get the `NodeRef` from the parent node, you can add an accessor (`functionRef` or such). Not exactly pretty, but we do that elsewhere already, and it'll go away with the new AST structure anyways.

Reply:
I think this should be fixed now, closing. Please feel free to reopen if you think this is still an issue.